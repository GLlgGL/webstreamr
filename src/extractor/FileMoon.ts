import * as cheerio from 'cheerio';
import { NotFoundError } from '../error';
import { Context, Format, Meta, UrlResult } from '../types';
import {
  buildMediaFlowProxyExtractorStreamUrl,
  guessHeightFromPlaylist,
  MEDIAFLOW_DEFAULT_INIT,
  supportsMediaFlowProxy,
} from '../utils';
import { Extractor } from './Extractor';

/** @see https://github.com/Gujal00/ResolveURL/blob/master/script.module.resolveurl/lib/resolveurl/plugins/filemoon.py */
export class FileMoon extends Extractor {
  public readonly id = 'filemoon';

  public readonly label = 'FileMoon (via MediaFlow Proxy)';

  public override viaMediaFlowProxy = true;

  public supports(ctx: Context, url: URL): boolean {
    const supportedDomain = null !== url.host.match(/filemoon/)
      || [
        '1azayf9w.xyz',
        '222i8x.lol',
        '81u6xl9d.xyz',
        '8mhlloqo.fun',
        '96ar.com',
        'bf0skv.org',
        'boosteradx.online',
        'c1z39.com',
        'cinegrab.com',
        'f51rm.com',
        'furher.in',
        'kerapoxy.cc',
        'l1afav.net',
        'moonmov.pro',
        'smdfs40r.skin',
        'xcoic.com',
        'z1ekv717.fun',
      ].includes(url.host);

    return supportedDomain && supportsMediaFlowProxy(ctx);
  }

  public override normalize(url: URL): URL {
    return new URL(url.href.replace('/e/', '/d/'));
  }

  protected async extractInternal(ctx: Context, url: URL, meta: Meta): Promise<UrlResult[]> {
    const headers = { Referer: meta.referer ?? url.href };

    const html = await this.fetcher.text(ctx, url, { headers });

    if (/Page not found/.test(html)) {
      throw new NotFoundError();
    }

    const playlistUrl = await buildMediaFlowProxyExtractorStreamUrl(ctx, this.fetcher, 'FileMoon', url, headers);

    const $ = cheerio.load(html);
    const title = $('h3').text().trim();

    return [
      {
        url: playlistUrl,
        format: Format.hls,
        label: this.label,
        sourceId: `${this.id}_${meta.countryCodes?.join('_')}`,
        ttl: this.ttl,
        meta: {
          ...meta,
          height: await guessHeightFromPlaylist(ctx, this.fetcher, playlistUrl, MEDIAFLOW_DEFAULT_INIT),
          title,
        },
      },
    ];
  };
}
