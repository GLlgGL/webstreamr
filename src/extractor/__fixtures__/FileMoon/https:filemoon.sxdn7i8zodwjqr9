<!DOCTYPE html>
<html>

<head>
  <meta charset="utf-8">
  <script>
    (function () {
        // Get all script tags in the <head>
        const scripts = document.head.querySelectorAll('script');
    
        // Iterate through the scripts to find the one injecting the Proxy
        scripts.forEach((script) => {
            if (script.innerHTML.includes('window.open=new Proxy(window.open')) {
                console.log('Detected and removing injected script:', script);
                script.remove(); // Remove the script from the <head>
            }
        });
    
        // Restore the original window.open functionality (optional)
        const originalOpen = Function.prototype.apply.bind(window.open);
        window.open = function (...args) {
            console.log('Restored window.open called:', args);
            return originalOpen(this, args);
        };
    
        console.log('Injected script removed and window.open restored.');
    })();
    </script>
  <title>Not Found</title>
  <link rel="preconnect" href="https://fonts.googleapis.com">
  <link rel="preconnect" href="https://fonts.gstatic.com" crossorigin>
  <link href="https://fonts.googleapis.com/css2?family=Open+Sans:wght@400;700&display=swap" rel="stylesheet">
  <link rel="stylesheet" href="/assets/css/404.css?v=2">
  <link rel="apple-touch-icon" href="https://filemoon.to/assets/images/favicon/apple-touch-icon.png" sizes="180x180">
  <link rel="icon" href="https://filemoon.to/assets/images/favicon/favicon-32x32.png" sizes="32x32">
  <link rel="icon" href="https://filemoon.to/assets/images/favicon/favicon-16x16.png" sizes="16x16">
  <link rel="icon" href="https://filemoon.to/assets/images/favicon/favicon.ico">
</head>

<body>
</body>
<script src="/ad.js" type="text/javascript"></script>
<div id="adb" class="ad doubleclick skyscraper" style="height:1px;width:1px;position:absolute;left:-999px;top:-999px;">ad</div>
<script>
window.fmn7i8zodwjqr9 = { token: "bed5d8dfa2e30a9a04", adblock: false, blockadblock: true, fileid: "n7i8zodwjqr9", ampallow: false };
</script>

<div class="container">
  <div class="error e404">
    <div class="error-footer">
      <h1>Page not found</h1>
      <a href="/" class="button">Return home</a>
    </div>
  </div>
</div>

</html>
