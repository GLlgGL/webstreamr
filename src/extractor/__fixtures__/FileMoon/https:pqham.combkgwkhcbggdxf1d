<HTML class="a-bean">
<HEAD>
  
  <link rel="apple-touch-icon" href="https://filemoon.to/assets/images/favicon/apple-touch-icon.png" sizes="180x180">
  <link rel="icon" href="https://filemoon.to/assets/images/favicon/favicon-32x32.png" sizes="32x32">
  <link rel="icon" href="https://filemoon.to/assets/images/favicon/favicon-16x16.png" sizes="16x16">
  <link rel="icon" href="https://filemoon.to/assets/images/favicon/favicon.ico">
  
  <meta name="robots" content="noindex, nofollow">

<link rel="stylesheet" href="/css/main.css?v=4">
<script type="text/javascript" src="/js/jquery.js"></script>
<script type="text/javascript" src="/js/xupload.js?v=3"></script>
<script type="text/javascript" src="/js/jquery.cookie.js"></script>
<script type="text/javascript" src='/js/ls.js'></script>
<script>
$.cookie('file_id', '48329866', { expires: 10 });
$.cookie('aff', '31148', { expires: 10 });
$.cookie('ref_url', 'pqham.com', { expires: 10 });
</script>
</HEAD>
<BODY topmargin=0 leftmargin=0 style="background:transparent;" class="service-tho">


  
	


	
	<script type="text/javascript" src="/js/pcap.js"></script>
	
		
			<script type="text/javascript">
				aclib.runPop({
					zoneId: '10489682',
				});
			</script>
		
		
	

	<!-- Anime -->
	

	<!-- Adult -->
	


	


<script>var Zordon = undefined, ZorDon = undefined;</script>
<script type="text/javascript" src="/js/bafsd.js" integrity="sha384-AZIiMnh/6CCMbnd5esCXv/4OwYHf+apmHobUAlFkLorsu+aMLR+nwSElHsqMoB8j"></script>
<script>
var a=function(){window.ZorDon=!0},b=function(){window.ZorDon=!1};"undefined"==typeof ZorDon?$(document).ready(a):Zordon.on(!0,a).on(!1,b),Zordon=void 0
</script>

<script>
!function(){try{var t=["sandbox","hasAttribute","frameElement","data","indexOf","href","domain","","plugins","undefined","namedItem","Chrome PDF Viewer","object","createElement","on" + "error","type","application/pdf","setAttribute","style","visibility:hidden;width:0;height:0;position:absolute;top:-99px;","data:application/pdf;base64,JVBERi0xLg0KdHJhaWxlcjw8L1Jvb3Q8PC9QYWdlczw8L0tpZHNbPDwvTWVkaWFCb3hbMCAwIDMgM10+Pl0+Pj4+Pj4=","appendChild","body","removeChild","parentElement","/blocked.html?referer=","substring","referrer"];function e(){try{if(config.ampallow){var e=window.location.ancestorOrigins;if(e[e.length-1].endsWith("ampproject.org"))return}}catch(n){}setTimeout(function(){location[t[5]]="/blocked.html"},900)}!function e(n){try{if(window[t[2]][t[1]](t[0])){n();return}}catch(r){}if(0!=location[t[5]][t[4]](t[3])&&document[t[6]]==t[7]){n();return}if(typeof navigator[t[8]]!=t[9]&&typeof navigator[t[8]][t[10]]!=t[9]&&null!=navigator[t[8]][t[10]](t[11])){var i=document[t[13]](t[12]);i[t[14]]=function(){n()},i[t[17]](t[15],t[16]),i[t[17]](t[18],t[19]),i[t[17]](t[3],t[20]),document[t[22]][t[21]](i),setTimeout(function(){i[t[24]][t[23]](i)},150)}}(e),function t(){try{document.domain=document.domain}catch(e){try{if(-1!=e.toString().toLowerCase().indexOf("sandbox"))return!0}catch(n){}}return!1}()&&e(),function t(){if(window.parent===window)return!1;try{var e=window.frameElement}catch(n){e=null}return null===e?""===document.domain&&"data:"!==location.protocol:e.hasAttribute("sandbox")}()&&e()}catch(n){}}();
</script>



    <div style="position:relative;" class="mb-4 mb-lg-5" id="iframe-holder">


    <script type='text/javascript' src='/player/jw8_26/jwplayer.js?v=5.0.2'></script>
				<script type="text/javascript">jwplayer.key="eNFaXCjyURVoCCGiHp7HTQ3hDhE/AfL0g8VE1fRbL84=";</script>
                <script src="/js/dnsads.js?dfp=1&ad_code=2&adsrc=3"></script>
                <div id='vplayer' class='vid-player' style="width:100%;height: 100%;text-align:center;"></div>


    </div>




<script data-cfasync='false' type='text/javascript'>eval(function(p,a,c,k,e,d){while(c--)if(k[c])p=p.replace(new RegExp('\\b'+c.toString(a)+'\\b','g'),k[c]);return p}('l d=u("1c");d.a3({a2:[{a1:"2z://a0.9z.9y.9x.4e/9w/9v/9u/9t/9s.9r?t=9q-9p&s=4f&e=9o&f=4g&9n=9m&9l=9k&9j=9i&p="}],9h:"2z://9g.9f/19.9e",2p:"1z%",2o:"1z%",9d:"9c",9b:\'9a\',99:"98",97:"96",q:[],"95":{"94":"93","92":"91","90":1b},\'8z\':{"8y":"8x"},8w:{2q:"/8v/1y/51-4z.1y?v=3.0.6",2c:"51-4z"},8u:"8t",8s:"2z://8r.8q",8p:{},8o:1b,8n:[0.25,0.50,0.75,1,1.25,1.5,2]});l 2w,2y,8m;l 8l=0,8k=0;l d=u("1c");l 4y=0,8j=0,8i=0,1n=0;l 2m=\'19\';$.8h({8g:{\'8f-8e\':\'46-8d\'}});d.k(\'3e\',9(x){8(5>0&&x.1d>=5&&2y!=1){2y=1;$(\'z.8c\').8b(\'8a\')}8(x.1d>=1n+5||x.1d<1n){1n=x.1d;2i.89(\'19\',2k.88(1n),{87:60*60*24})}});d.k(\'1i\',9(x){4y=x.1d});d.k(\'1s\',9(x){2x(x);$(\'.7-18-41\').2l()});d.k(\'86\',9(){$(\'z.4w\').85();2i.2l(\'19\')});$(4k).k(\'y\',9(17){o 1l=17.84-$(\'#1c\').4x().48;o 1k=17.83-$(\'#1c\').4x().2n;2x(1l,1k)});9 2x(1l,1k){$(\'z.4w\').3f();8(2w)1p;2w=1;o 21=0;8(2t.82===1b){21=1}o 14="26";8(/81|4u/i.j(g.h)){14=/80|7z|7y/i.j(g.h)?"4v/7x":"4v/4u"}r 8(/4t/i.j(g.h)){14="26/4t"}r 8(/4s/i.j(g.h)){14="26/4s"}r 8(/7w|7v 7u 7t/i.j(g.h)){14="26/7s"}r 8(/4r/i.j(g.h)){14="4q/4r"}r 8(/4p/i.j(g.h)){14="4q/4p"}r 8(/7r|7q|7p|7o/i.j(g.h)){14="7n"}o 11="7m";8(/4o/i.j(g.h)){11="4o"}r 8(/2v/i.j(g.h)&&!/2u|4l/i.j(g.h)){11="2v"}r 8(/4n/i.j(g.h)&&!/2v/i.j(g.h)){11="4n"}r 8(/4m/i.j(g.h)){11="4m"}r 8(/2u/i.j(g.h)){11="2u"}r 8(/4l/i.j(g.h)){11="7l"}r 8(/7k|7j/i.j(g.h)){11="7i 7h"}8(g.7g){11+=" (7f)"}o 4d=2t.7e;o 4c=2t.7d;o 2s="";o 2r="";o 4b=g.7c;7b{o 23=4k.7a(\'23\');o 1m=23.4j(\'4i\')||23.4j(\'79-4i\');8(1m){o 22=1m.78(\'77\');8(22){2s=1m.4h(22.76)||"";2r=1m.4h(22.74)||""}}}73(e){}o 2q=\'/72?b=71&2m=19&70=4g-6z-28-4f-6y&6x=1&6w=&6v=6u.4e&6t=0&21=\'+21+\'&1l=\'+1l+\'&1k=\'+1k+\'&6s=\'+20(14)+\'&6r=\'+20(11)+\'&6q=\'+4d+\'&6p=\'+4c+\'&6o=\'+20(2s)+\'&6n=\'+20(2r)+\'&6m=\'+4b;$.3k(2q,9(4a){$(\'#6l\').6k(4a)});$(\'.7-c-6j-6i:6h("6g")\').y(9(e){49();u().6f(0);u().6e(1b)});9 49(){l $1x=$("<z />").1y({1d:"6d",2p:"1z%",2o:"1z%",48:0,2n:0,47:6c,6b:"6a(10%, 10%, 10%, 0.4)","2g-69":"68"});$("<67 />").1y({2p:"60%",2o:"60%",47:66,"65-2n":"64"}).63({\'62\':\'/?b=61&2m=19\',\'5z\':\'0\',\'5y\':\'46\'}).45($1x);$1x.y(9(){$(5x).2l();u().1s()});$1x.45($(\'#1c\'))}u().1i(0)}9 42(2j){l 43=2k.44(2j/60);l 1j=2k.44(2j%60);8(1j<10){1j="0"+1j}1p 43+":"+1j}9 3j(a){a=42(a);$(\'#1c\').3d(`<z 1w="7-18-41"><z 1w="7-18-2g">5w 5v ${a}</z><1v 1w="7-18-13">5u</1v><1v 1w="7-18-3i">5t</1v></z>`)}9 5s(){l q=d.29(40);3z.3y(q);8(q.1g>1){32(i=0;i<q.1g;i++){8(q[i].2c==40){3z.3y(\'!!=\'+i);d.30(i)}}}}d.k(\'5r\',9(){u().2e(\'<1h 3x="3w://3v.3u.3t/3s/1h" 3r:3q="3p" 3o=" 0 0 1 1" 3n="1.1"/>\',"5q 10 3m",9(){u().1i(u().3l()+10)},"1r");$("z[13=1r]").3h().3g(\'.7-w-2h\');u().2e(\'<1h 3x="3w://3v.3u.3t/3s/1h" 3r:3q="3p" 3o=" 0 0 1 1" 3n="1.1"/>\',"5p 10 3m",9(){l 1u=u().3l()-10;8(1u<0)1u=0;u().1i(1u)},"1q");l 1t=2i.3k(\'19\');8(1t!==5o){3j(1t);$(\'1f\').k(\'y\',\'.7-18-13\',9(){d.1s();35(9(){d.1i(1t)},5n)});$(\'1f\').k(\'y\',\'.7-18-3i\',9(){u().1s()})}$("z[13=1q]").3h().3g(\'.7-w-2h\');$("z.7-w-2h").3f();$(\'.7-5m-3e\').3d($(\'.7-2g-5l\'));$(\'1f\').k(\'y\',\'.7-3c-w-3b .7-w[13="1r"]\',9(){$(\'.7-2d .7-w[13="1r"]\').3a(\'y\')});$(\'1f\').k(\'y\',\'.7-3c-w-3b .7-w[13="1q"]\',9(){$(\'.7-2d .7-w[13="1q"]\').3a(\'y\')})});9 2f(){}d.k(\'5k\',9(){2f()});d.k(\'5j\',9(){2f()});d.k("m",9(17){l q=d.29();8(q.1g<2)1p;$(\'.7-c-5i-5h\').5g(9(){$(\'#7-c-n-m\').1o(\'7-c-n-1e\');$(\'.7-n-m\').16(\'12-15\',\'1a\')});d.2e("/5f/5e.1h","37 36",9(e){$(\'.7-39\').5d(\'7-c-38\');8($(\'.7-39\').5c(\'7-c-38\')){$(\'.7-c-m\').16(\'12-15\',\'1b\');$(\'.7-c-n-m \').16(\'12-15\',\'1b\');$(\'.7-c-n-m \').5b(\'7-c-n-1e\')}r{$(\'.7-c-m\').16(\'12-15\',\'1a\');$(\'.7-c-n-m \').16(\'12-15\',\'1a\');$(\'.7-c-n-m \').1o(\'7-c-n-1e\')}$(\'.7-2d .7-w:5a([12-59="37 36"])\').k(\'y\',9(){$(\'.7-c-m\').16(\'12-15\',\'1a\');$(\'.7-c-n-m \').16(\'12-15\',\'1a\');$(\'.7-c-n-m \').1o(\'7-c-n-1e\')})},"58");d.k("57",9(17){2b.56(\'2a\',17.q[17.55].2c)});8(2b.34(\'2a\')){35("33(2b.34(\'2a\'));",54)}});l 27;9 33(31){l q=d.29();8(q.1g>1){32(i=0;i<q.1g;i++){8(q[i].53==31){8(i==27){1p}27=i;d.30(i)}}}}$(\'1f\').k(\'y\',\'.7-w-c\',9(){$(\'.7-c-n-m \').1o(\'7-c-n-1e\');$(\'.7-13-52.7-c-m\').16(\'12-15\',\'1a\')});',36,364,'|||||||jw|if|function|||settings|videop|||navigator|userAgent||test|on|var|audioTracks|submenu|let||tracks|else|||jwplayer||icon||click|div||browserName|aria|button|deviceType|expanded|attr|event|resume|wkhcbggdxf1d|false|true|vplayer|position|active|body|length|svg|seek|remainingSeconds|cy|cx|gl|lastt|removeClass|return|ff00|ff11|play|savedTime|tt|span|class|dd|css|100|encodeURIComponent|adb|debugInfo|canvas|||Desktop|current_audio||getAudioTracks|default_audio|localStorage|name|controlbar|addButton|callMeMaybe|text|rewind|ls|seconds|Math|remove|file_code|top|height|width|url|gpuRenderer|gpuVendor|window|Edge|Chrome|vvplay|doPlay|vvad|https|setCurrentAudioTrack|audio_iso|for|audio_set|getItem|setTimeout|Track|Audio|open|controls|trigger|container|display|append|time|hide|insertAfter|detach|reset|addResume|get|getPosition|sec|version|viewBox|preserve|space|xml|2000|org|w3|www|http|xmlns|log|console|track_name|box|formatTime|minutes|floor|appendTo|no|zIndex|left|showCCform|data|metaxa|windowHeight|windowWidth|com|1759839668|48329866|getParameter|webgl|getContext|document|OPR|Firefox|Safari|Brave|PlayStation|Console|Xbox|Linux|Windows|Android|Mobile|video_ad|offset|prevt|theme||jw8|color|language|300|currentTrack|setItem|audioTrackChanged|dualSound|label|not|addClass|hasClass|toggleClass|dualy|images|mousedown|buttons|topbar|playAttemptFailed|beforePlay|countdown|slider|1500|null|Rewind|Forward|ready|set_audio_track|No|Yes|at|Resume|this|scrolling|frameborder||upload_srt|src|prop|50px|margin|1000001|iframe|center|align|rgba|background|1000000|absolute|pause|setCurrentCaptions|Upload|contains|item|content|html|fviews|mk|gpur|gpuv|wh|ww|browser|device|vb|pqham|referer|prem|embed|85da50da22accdb23c12080afdf09544|104|hash|view|dl|catch|UNMASKED_RENDERER_WEBGL||UNMASKED_VENDOR_WEBGL|WEBGL_debug_renderer_info|getExtension|experimental|createElement|try|hardwareConcurrency|innerHeight|innerWidth|Selenium|webdriver|Explorer|Internet|Trident|MSIE|Opera|Unknown|TV|NetCast|Web0S|Tizen|SmartTV|MacOS|X|OS|Mac|Macintosh|iOS|iPod|iPad|iPhone|Mobi|ZorDon|pageY|pageX|show|complete|ttl|round|set|slow|fadeIn|video_ad_fadein|cache|Cache|Content|headers|ajaxSetup|v2done|tott|vastdone2|vastdone1|vvbefore|playbackRates|playbackRateControls|cast|sx|filemoon|aboutlink|FileMoon|abouttext|assets|skin|720p|852|qualityLabels|preloadAds|insecure|vpaidmode|vast|client|advertising|start|startparam|none|fullscreenOrientationLock|metadata|preload|uniform|stretching|jpg|me|videothumbs|image|4000|sp|13335|asn|1075|srv|10800|4bLs9AVOmc|ZS23QrFqZGvmZWBThghesV8ebtIIW8qX|m3u8|master|wkhcbggdxf1d_h|09665|01|hls2|i8yz83pn|waw05|rcr82|be7713|file|sources|setup'.split('|')))
</script>
</BODY></HTML>
