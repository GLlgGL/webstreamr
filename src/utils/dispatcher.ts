import { socksDispatcher } from 'fetch-socks';
import { Agent, Dispatcher, interceptors, ProxyAgent } from 'undici';

const createBasicDispatcher = (): Dispatcher => {
  if (process.env['ALL_PROXY']) {
    const proxyUrl = new URL(process.env['ALL_PROXY']);
    if (proxyUrl.protocol === 'socks5:') {
      return socksDispatcher({ type: 5, host: proxyUrl.hostname, port: parseInt(proxyUrl.port) }, { allowH2: true });
    } else {
      return new ProxyAgent({ uri: proxyUrl.href, allowH2: true });
    }
  } else {
    return new Agent({ allowH2: true });
  }
};

export const createDispatcher = (): Dispatcher => {
  return createBasicDispatcher().compose(
    interceptors.dns(),
    interceptors.retry({ maxRetries: 3 }),
  );
};
