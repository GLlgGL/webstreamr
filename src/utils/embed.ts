import { unpack } from 'unpacker';

export const unpackEval = (html: string): string => {
  const evalMatch = html.match(/eval\(function\(p,a,c,k,e,d\).*\)\)/);
  if (!evalMatch) {
    throw new Error(`No p.a.c.k.e.d string found`);
  }

  return unpack(evalMatch[0]);
};

export const extractUrlFromPacked = (html: string, linkRegExps: RegExp[]): URL => {
  const unpacked = unpackEval(html);

  for (const linkRegexp of linkRegExps) {
    const linkMatch = unpacked.match(linkRegexp);
    if (linkMatch && linkMatch[1]) {
      return new URL(`https://${linkMatch[1].replace(/^(https:)?\/\//, '')}`);
    }
  }

  throw new Error(`Could not find a stream link in embed`);
};
