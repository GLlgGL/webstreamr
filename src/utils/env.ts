import * as os from 'node:os';
import { Request } from 'express';

export const envGet = (name: string): string | undefined => process.env[name];

export const envGetRequired = (name: string): string => {
  const value = envGet(name);
  if (!value) {
    throw new Error(`Environment variable "${name}" is not configured.`);
  }

  return value;
};

export const envGetAppId = (): string => process.env['MANIFEST_ID'] || 'webstreamr';

export const envGetAppName = (): string => process.env['MANIFEST_NAME'] || 'WebStreamr';

export const envIsProd = (): boolean => process.env['NODE_ENV'] === 'production';

export const isElfHostedInstance = (req: Request): boolean => req.host.endsWith('elfhosted.com');

export const getCacheDir = (): string => envGet('CACHE_DIR') ?? os.tmpdir();
