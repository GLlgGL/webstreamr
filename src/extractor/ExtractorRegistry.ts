import TTLCache from '@isaacs/ttlcache';
import winston from 'winston';
import { Context, CountryCode, UrlResult } from '../types';
import { isExtractorDisabled } from '../utils';
import { Extractor } from './Extractor';

export class ExtractorRegistry {
  private readonly logger: winston.Logger;
  private readonly extractors: Extractor[];
  private readonly urlResultCache: TTLCache<string, UrlResult[]>;

  public constructor(logger: winston.Logger, extractors: Extractor[]) {
    this.logger = logger;
    this.extractors = extractors;
    this.urlResultCache = new TTLCache({ max: 4096 });
  }

  public async handle(ctx: Context, url: URL, countryCode: CountryCode, title?: string | undefined): Promise<UrlResult[]> {
    const extractor = this.extractors.find(extractor => !isExtractorDisabled(ctx.config, extractor) && extractor.supports(ctx, url));
    if (!extractor) {
      return [];
    }

    const normalizedUrl = extractor.normalize(url);
    const cacheKey = extractor.viaMediaFlowProxy
      ? `${extractor.id}_${normalizedUrl}_${ctx.config.mediaFlowProxyUrl}`
      : `${extractor.id}_${normalizedUrl}`;

    let urlResults = this.urlResultCache.get(cacheKey) ?? [];
    if (this.urlResultCache.has(cacheKey)) {
      return urlResults.map(urlResult => ({ ...urlResult, ttl: this.urlResultCache.getRemainingTTL(cacheKey) }));
    }

    this.logger.info(`Extract stream URL using ${extractor.id} extractor from ${url}`, ctx);

    urlResults = await extractor.extract(ctx, normalizedUrl, countryCode, title);

    if (!urlResults.some(urlResult => urlResult.error) && extractor.ttl) {
      this.urlResultCache.set(cacheKey, urlResults, { ttl: extractor.ttl });
    }

    return urlResults;
  };
}
