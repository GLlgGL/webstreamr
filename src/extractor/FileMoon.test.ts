import winston from 'winston';
import { createTestContext } from '../test';
import { FetcherMock } from '../utils';
import { ExtractorRegistry } from './ExtractorRegistry';
import { FileMoon } from './FileMoon';

const logger = winston.createLogger({ transports: [new winston.transports.Console({ level: 'nope' })] });
const extractorRegistry = new ExtractorRegistry(logger, [new FileMoon(new FetcherMock(`${__dirname}/__fixtures__/FileMoon`))]);

const ctx = createTestContext({ mediaFlowProxyUrl: 'https://mediaflow.test.org', mediaFlowProxyPassword: 'test' });

describe('FileMoon', () => {
  test('z1ekv717 d', async () => {
    expect(await extractorRegistry.handle(ctx, new URL('https://z1ekv717.fun/d/wkhcbggdxf1d'))).toMatchSnapshot();
  });

  test('page not found', async () => {
    expect(await extractorRegistry.handle(ctx, new URL('https://filemoon.sx/e/n7i8zodwjqr9'))).toMatchSnapshot();
  });
});
