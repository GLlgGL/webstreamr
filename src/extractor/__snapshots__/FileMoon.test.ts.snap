// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`FileMoon page not found 1`] = `[]`;

exports[`FileMoon z1ekv717 d 1`] = `
[
  {
    "format": "hls",
    "label": "FileMoon (via MediaFlow Proxy)",
    "meta": {
      "countryCodes": [],
      "height": 720,
      "title": "alien earth s01e01 1080p web h264-successfulcrab",
    },
    "sourceId": "filemoon_",
    "ttl": 900000,
    "url": "https://mediaflow.test.org/proxy/hls/manifest.m3u8?api_password=test&h_user-agent=Mozilla%2F5.0+%28Windows+NT+10.0%3B+Win64%3B+x64%29+AppleWebKit%2F537.36+%28KHTML%2C+like+Gecko%29+Chrome%2F136.0.0.0+Safari%2F537.36&h_referer=https%3A%2F%2Fz1ekv717.fun%2Fd%2Fwkhcbggdxf1d&d=https%3A%2F%2Fbe7713.rcr82.waw05.i8yz83pn.com%2Fhls2%2F01%2F09665%2Fwkhcbggdxf1d_h%2Fmaster.m3u8%3Ft%3D359izbQ_L-sOHaIjj275YKdBDyiPzUi3ETfbNoN9bzU%26s%3D1759754698%26e%3D10800%26f%3D48329866%26srv%3D1075%26asn%3D13335%26sp%3D4000%26p%3D",
  },
]
`;
